### PR TITLE
[RFR] Introduce useGetMatching and useReferenceArrayInputController

### PR DESCRIPTION
**Pull request by djhi**
- [x] useGetMatching
- [x] useReferenceArrayInputController
- [x] Fix tests

### CODE REVIEW

File: packages/ra-core/src/controller/input/ReferenceArrayInputController.spec.tsx [hunk 1]
```diff
@@ -198,34 +205,15 @@ describe('<ReferenceArrayInputController />', () => {
...
-            >
-                {children}
-            </ReferenceArrayInputController>
-        );
-        assert.equal(children.mock.calls[0][0].onChange, onChange);
+        expect(children.mock.calls[0][0].warning).toEqual(undefined);
     });
 
-    it('should call crudGetMatching on mount with default fetch values', () => {
+    it.skip('should call crudGetMatching on mount with default fetch values', () => {
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/input/ReferenceArrayInputController.spec.tsx]: skip detected, here and down under

File: packages/ra-core/src/controller/input/ReferenceArrayInputController.spec.tsx [hunk 2]
```diff
@@ -198,34 +205,15 @@ describe('<ReferenceArrayInputController />', () => {
                 {children}
             </ReferenceArrayInputController>
         );
-        assert.equal(children.mock.calls[0][0].warning, undefined);
-    });
-
-    it('should pass onChange down to child component', () => {
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/input/ReferenceArrayInputController.spec.tsx]: why did you remove this one?

File: packages/ra-core/src/controller/input/ReferenceArrayInputController.spec.tsx [hunk 3]
```diff
@@ -269,207 +256,181 @@ describe('<ReferenceArrayInputController />', () => {
...
-            },
-            {
-                field: 'id',
-                order: 'DESC',
-            },
-            {
-                q: 'bar',
-            },
-        ]);
+        // wrapper.instance().setFilter('bar');
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/input/ReferenceArrayInputController.spec.tsx]: commented out code (in a skipped test). You *really* don't want that code to run!

File: packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx [hunk 4]
```diff
@@ -1,34 +1,19 @@
...
-    `${resource}@${source}`;
+import { Record, Sort, Pagination } from '../../types';
+import useReferenceArrayInputController from './useReferenceArrayInputController';
 
 interface ChildrenFuncParams {
     choices: Record[];
     error?: string;
+    loaded: boolean;
     loading: boolean;
-    onChange: (value: any) => void;
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx]: I still don't understand why the new controller doesn't return this one.

File: packages/ra-core/src/controller/input/useReferenceArrayInputController.ts [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+ * @param {string} option.reference The linked resource name
+ * @param {string} option.resource The current resource name
+ * @param {string} option.source The key of the linked resource identifier
+ *
+ * @return {Object} controllerProps Fetched data and callbacks for the ReferenceArrayInput components
+ */
+const useReferenceArrayInputController = ({
+    basePath,
+    filter: defaultFilter,
+    filterToQuery = searchText => ({ q: searchText }),
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/input/useReferenceArrayInputController.ts]: Move the default value to a constant to avoid having a new function at every render, defeating your useMemo line 93

File: packages/ra-core/src/controller/input/useReferenceArrayInputController.ts [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    options,
+    reference,
+    resource,
+    source,
+}: Option): ReferenceArrayInputProps => {
+    const translate = useTranslate();
+
+    const [pagination, setPagination] = useState({ page: 1, perPage });
+    const [sort, setSort] = useState(defaultSort);
+    const [filter, setFilter] = useState('');
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/input/useReferenceArrayInputController.ts]: the default filter should be an empty object

File: packages/ra-core/src/controller/input/useReferenceArrayInputController.ts [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        reference,
+        pagination,
+        sort,
+        finalFilter,
+        source,
+        resource,
+        options
+    );
+
+    const finalReferenceRecords = referenceRecords.filter(
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/input/useReferenceArrayInputController.ts]: We need a comment here, like "filter out not found references - happens when the dataProvider doesn't guarantee referential integrity"

File: packages/ra-core/src/controller/input/useReferenceArrayInputController.ts [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    const finalMatchingReferences =
+        matchingReferences && matchingReferences.length > 0
+            ? (matchingReferences || []).concat(finalReferenceRecords)
+            : finalReferenceRecords.length > 0
+            ? finalReferenceRecords
+            : undefined;
+
+    const dataStatus = getDataStatus({
+        input,
+        // We merge the currently selected records with the matching ones, otherwise
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/input/useReferenceArrayInputController.ts]: I think you should move this comment up to line 115

File: packages/ra-core/src/dataProvider/useGetMatching.ts [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+import {
+    getReferenceResource,
+    getPossibleReferenceValues,
+    getPossibleReferences,
+} from '../reducer';
+
+const referenceSource = (resource, source) => `${resource}@${source}`;
+
+/**
+ * Call the dataProvider with a GET_LIST verb and return the result as well as the loading state.
```
Comment (fzaninotto) [inline packages/ra-core/src/dataProvider/useGetMatching.ts]: The comment is the same as useGetList... You should tweak it so that readers understand the difference

File: packages/ra-core/src/dataProvider/useGetMatching.ts [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+ * @param {string} referencingResource The resource name, e.g. 'posts'. Used to build a cache key
+ * @param {Object} options Options object to pass to the dataProvider. May include side effects to be executed upon success of failure, e.g. { onSuccess: { refresh: true } }
+ *
+ * @returns The current request state. Destructure as { data, total, ids, error, loading, loaded }.
+ *
+ * @example
+ *
+ * import { useGetMatching } from 'react-admin';
+ *
+ * const LatestNews = () => {
```
Comment (fzaninotto) [inline packages/ra-core/src/dataProvider/useGetMatching.ts]: Bad example

File: packages/ra-core/src/controller/input/ReferenceArrayInputController.spec.tsx (hunk 1 above)
Comment (djhi) [reply packages/ra-core/src/controller/input/ReferenceArrayInputController.spec.tsx]: Yes, because this PR is WIP ;) It event has a `Fix tests` in its description!

File: packages/ra-core/src/controller/input/ReferenceArrayInputController.spec.tsx (hunk 2 above)
Comment (djhi) [reply packages/ra-core/src/controller/input/ReferenceArrayInputController.spec.tsx]: Because onChange override is handled at the input level. See the `useInput` hook

File: packages/ra-core/src/dataProvider/useGetMatching.ts (hunk 9 above)
Comment (djhi) [reply packages/ra-core/src/dataProvider/useGetMatching.ts]: It is tweaked in the example. GET_MANY is actually sent as GET_LIST

File: packages/ra-core/src/controller/input/useReferenceArrayInputController.ts (hunk 6 above)
Comment (djhi) [reply packages/ra-core/src/controller/input/useReferenceArrayInputController.ts]: No, because it goes through `filterToQuery` which then transform it into an object

File: packages/ra-core/src/controller/input/ReferenceArrayInputController.spec.tsx [hunk 11]
```diff
@@ -1,475 +1,633 @@
...
+        expect(dispatch).toHaveBeenCalledWith({
+            type: CRUD_GET_MANY,
+            meta: {
+                resource: 'tags',
+            },
+            payload: { ids: [6] },
+        });
     });
 
-    it('should only call crudGetOne and not crudGetMatching when only the record changes', () => {
```
Comment (djhi) [inline packages/ra-core/src/controller/input/ReferenceArrayInputController.spec.tsx]: I removed this test as the component should not care about record, only about the value

File: packages/ra-core/src/controller/input/ReferenceArrayInputController.spec.tsx [hunk 12]
```diff
@@ -1,475 +1,633 @@
...
-            'tags',
-            'posts@tag_ids',
-            { page: 1, perPage: 42 },
-            { field: 'foo', order: 'ASC' },
-            { foo: 'bar' },
-        ]);
-        assert.equal(crudGetMany.mock.calls.length, 1);
-    });
 
-    it('should call crudGetMany when input value changes', () => {
```
Comment (djhi) [inline packages/ra-core/src/controller/input/ReferenceArrayInputController.spec.tsx]: Merged this test and the next one

File: packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx (hunk 4 above)
Comment (djhi) [reply packages/ra-core/src/controller/input/ReferenceArrayInputController.tsx]: > Because `onChange` override is handled at the input level. See the `useInput` hook
